In [83]:
# Import Libraries

In [84]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [85]:
# import ords_prods_merged

In [86]:
ords_prods_merged = pd.read_pickle(r'/Users/pedropaulo/Desktop/Tasks CF/Instacart Basket Analysis/Data/Prepared Data/orders_products_merged_3.pkl')

In [87]:
# import df_customer

In [88]:
df_customer = pd.read_csv(r'/Users/pedropaulo/Desktop/Tasks CF/Instacart Basket Analysis/Data/Original Data/customers.csv')

In [89]:
# Checking customer

In [90]:
df_customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [91]:
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [92]:
# Checking ords_prods_merged

In [93]:
ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...  _merge     busiest_day    busiest_days  \
0         Soda  ...    both  Regularly busy  Regularly days   
1         Soda  ...    both  Regularly busy    Slowest days   
2         Soda  ...    both  Regularly busy    Slowest days   
3         Soda  ...    both      Least busy    Slowest days   
4         Soda  ...    both      Least busy    Slowest days   

  busiest_period_of_day max_order  loyalty_flag mean_price  spending_flag  \
0        Average orders        10  New customer   6.367797    Low spender   
1        Average orders        10  New customer   6.367797    Low spender   
2           Most orders        10  New customer   6.367797    Low spender   
3        Average orders        10  New customer   6.367797    Low spender   
4           Most orders        10  New customer   6.367797    Low spender   

  median_days         frequency_flag  
0        20.5  Non-frequent customer  
1        20.5  Non-frequent customer  
2        20.5  Non-frequent customer  
3        20.5  Non-frequent customer  
4        20.5  Non-frequent customer  

[5 rows x 23 columns]

In [94]:
#delete merge flag for next merge
del ords_prods_merged['_merge']

In [95]:
ords_prods_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000000 entries, 0 to 999999
Data columns (total 22 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   order_id               1000000 non-null  int64  
 1   user_id                1000000 non-null  int64  
 2   order_number           1000000 non-null  int64  
 3   orders_day_of_week     1000000 non-null  int64  
 4   order_hour_of_day      1000000 non-null  int64  
 5   days_since_last_order  940681 non-null   float64
 6   product_id             1000000 non-null  int64  
 7   add_to_cart_order      1000000 non-null  int64  
 8   reordered              1000000 non-null  int64  
 9   product_name           1000000 non-null  object 
 10  aisle_id               1000000 non-null  float64
 11  department_id          1000000 non-null  float64
 12  prices                 1000000 non-null  float64
 13  busiest_day            1000000 non-null  object 
 14  busiest_days       

In [96]:
# Arranging columns in the dataframe customers

In [97]:
df_customer = df_customer.rename(columns={'First Name':'first_name', 'Surnam':'last_name', 'Gender':'sex', 'Age':'age','STATE':'state'})


In [98]:
df_customer.head()

user_id first_name last_name     sex       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [99]:
# Checking Dataframes 

In [100]:
df_customer.dtypes

user_id          int64
first_name      object
last_name       object
sex             object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [101]:
ords_prods_merged.dtypes

order_id                   int64
user_id                    int64
order_number               int64
orders_day_of_week         int64
order_hour_of_day          int64
days_since_last_order    float64
product_id                 int64
add_to_cart_order          int64
reordered                  int64
product_name              object
aisle_id                 float64
department_id            float64
prices                   float64
busiest_day               object
busiest_days              object
busiest_period_of_day     object
max_order                  int64
loyalty_flag              object
mean_price               float64
spending_flag             object
median_days              float64
frequency_flag            object
dtype: object

In [102]:
# Checking missing values and duplicates in Customers Dataframe

In [103]:
df_customer.shape

(206209, 10)

In [104]:
df_customer.isnull().sum()

user_id             0
first_name      11259
last_name           0
sex                 0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [105]:
df_dups = df_customer[df_customer.duplicated()]

In [106]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, sex, state, age, date_joined, n_dependants, fam_status, income]
Index: []

No duplicates was found

In [107]:
# handling missing data

In [108]:
df_customer['first_name'].value_counts(dropna = False)

NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: first_name, Length: 208, dtype: int64

In [109]:
df_customer.fillna('not available')

user_id first_name last_name     sex           state  age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [110]:
df_customer_clean = df_customer.fillna('not available')

In [111]:
df_customer_clean['first_name'].value_counts(dropna = False)

not available    11259
Marilyn           2213
Barbara           2154
Todd              2113
Jeremy            2104
                 ...  
Merry              197
Eugene             197
Garry              191
Ned                186
David              186
Name: first_name, Length: 208, dtype: int64

No duplicate data was found, but 11259 data were missing from the first_name column, that was replaced to 'not available'. Probably it was a data collection error.

In [112]:
# Merge Customer_clean and ords_prods dataframes on user_id

In [113]:
df_merge = df_customer_clean.merge(ords_prods_merged, on = 'user_id', indicator = True)

In [114]:
df_merge.head()

user_id first_name last_name     sex     state  age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   n_dependants fam_status  income  ...     busiest_day    busiest_days  \
0             3    married  165665  ...  Regularly busy    Busiest days   
1             3    married  165665  ...  Regularly busy  Regularly days   
2             3    married  165665  ...  Regularly busy    Busiest days   
3             3    married  165665  ...  Regularly busy  Regularly days   
4             3    married  165665  ...  Regularly busy    Slowest days   

   busiest_period_of_day  max_order  loyalty_flag  mean_price  spending_flag  \
0            Most orders          8  New customer        11.4   High spender   
1            Most orders          8  New customer        11.4   High spender   
2            Most orders          8  New customer        11.4   High spender   
3            Most orders          8  New customer        11.4   High spender   
4            Most orders          8  New customer        11.4   High spender   

   median_days    frequency_flag  _merge  
0         14.0  Regular customer    both  
1         14.0  Regular customer    both  
2         14.0  Regular customer    both  
3         14.0  Regular customer    both  
4         14.0  Regular customer    both  

[5 rows x 32 columns]

In [115]:
df_merge['_merge'].value_counts()

both          1000000
left_only           0
right_only          0
Name: _merge, dtype: int64

In [116]:
# Export data as Pickle

In [117]:
df_merge.to_pickle(r'/Users/pedropaulo/Desktop/Tasks CF/Instacart Basket Analysis/Data/Prepared Data/orders_products_customers.pkl')